keras_cv/layers/object_detection/rpn_label_encoder.py

```diff
@@ -0,0 +1,176 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Mapping
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+from keras_cv.bounding_box import iou
+from keras_cv.ops import box_matcher
+from keras_cv.ops import sampling
+from keras_cv.ops import target_gather
+
+
+class _RpnLabelEncoder(tf.keras.layers.Layer):
+    """Transforms the raw labels into training targets for region proposal network (RPN).
+
+    # TODO(tanzhenyu): consider unifying with _ROISampler.
+    This is different from _ROISampler for a couple of reasons:
+    1) This deals with unbatched input, dict of anchors and potentially ragged labels
+    2) This deals with gt boxes, while _ROISampler deals with padded gt
+       boxes with value -1 and padded gt classes with value -1
+    3) this returns positive class target as 1, while _ROISampler returns
+       positive class target as-is. (All negative class target are 0)
+       The final classification loss will use one hot and #num_fg_classes + 1
+    4) this returns #num_anchors dense targets, while _ROISampler returns
+       #num_sampled_rois dense targets.
+    5) this returns all positive box targets, while _ROISampler still samples
+       positive box targets, while all negative box targets are also ignored
+       in regression loss.
+
+    Args:
+      anchor_format: The format of bounding boxes for anchors to generate. Refer
+        [to the keras.io docs](https://keras.io/api/keras_cv/bounding_box/formats/)
+        for more details on supported bounding box formats.
+      gt_box_format: The format of bounding boxes for ground truth boxes to generate.
+      positive_threshold: the threshold to set an anchor to positive match to gt box.
+        values above it are positive matches.
+      negative_threshold: the threshold to set an anchor to negative match to gt box.
+        values below it are negative matches.
+      samples_per_image: for each image, the number of positive and negative samples
+        to generate.
+      positive_fraction: the fraction of positive samples to the total samples.
+
+    """
+
+    def __init__(
+        self,
+        anchor_format,
+        gt_box_format,
+        positive_threshold,
+        negative_threshold,
+        samples_per_image,
+        positive_fraction,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.anchor_format = anchor_format
+        self.gt_box_format = gt_box_format
+        self.positive_threshold = positive_threshold
+        self.negative_threshold = negative_threshold
+        self.samples_per_image = samples_per_image
+        self.positive_fraction = positive_fraction
+        self.box_matcher = box_matcher.ArgmaxBoxMatcher(
+            thresholds=[negative_threshold, positive_threshold],
+            match_values=[-1, -2, 1],
```

Comment:
humm, why -2 for value not belong to either negative or positive?

Reply:
-2 is for things that will be ignored during training. This is documented in the box_matcher.py